lib/Browser.js

```diff
@@ -30,6 +30,55 @@ class Browser extends EventEmitter {
     this._screenshotTaskQueue = new TaskQueue();
     this._connection = connection;
     this._closeCallback = closeCallback || new Function();
+    /** @type {Map<string, Promise<Page>>} */
+    this._pages = new Map();
+    this._connection.on('Target.targetCreated', this._targetCreated.bind(this));
+    this._connection.on('Target.targetDestroyed', this._targetDestroyed.bind(this));
+  }
+
+  /**
+   * @param {!Connection} connection
+   * @param {boolean} ignoreHTTPSErrors
+   * @param {function()=} closeCallback
+   */
+  static async create(connection, ignoreHTTPSErrors, closeCallback) {
+    const browser =  new Browser(connection, ignoreHTTPSErrors, closeCallback);
+    await connection.send('Target.setDiscoverTargets', {discover: true});
+    return browser;
+  }
+
+  /**
+   * @param {{targetInfo: {targetId: string}}} param0
+   */
+  async _targetCreated({targetInfo}) {
+    if (targetInfo.type !== 'page' && targetInfo.type !== 'other')
+      return;
+
+    const page = await this._pageForTargetId(targetInfo.targetId);
+    this.emit(Browser.Events.PageOpened, page);
+  }
+
+  /**
+   * @param {{targetId: string}} param0
+   */
+  async _targetDestroyed({targetId}) {
+    const page = this._pages.get(targetId);
+    if (!page)
+      return;
+    this._pages.delete(targetId);
+    this.emit(Browser.Events.PageClosed, await page);
+  }
+
+  /**
+   * @param {string} targetId
+   * @return {Page}
+   */
+  _pageForTargetId(targetId) {
+    if (this._pages.has(targetId))
+      return this._pages.get(targetId);
+    const pagePromise = this._connection.createSession(targetId).then(client => Page.create(client, this._ignoreHTTPSErrors, this._screenshotTaskQueue));
```

Comment:
I don't think proactively attaching to all the targets is necessary. I'd rather expose the API as it is in the protocol - you can list the targets and you can materialize them as pages...